resolver/resolver.go

```diff
@@ -0,0 +1,92 @@
+		txCheckerRequests: requests,
+		multiStore:        multiStore,
+		cacher:            cacher,
+	}
+}
+
+func (resolver *Resolver) QueryBlock(ctx context.Context, id interface{}, params *jsonrpc.ParamsQueryBlock, req *http.Request) jsonrpc.Response {
+	return resolver.handleMessage(ctx, id, jsonrpc.MethodQueryBlock, params, req, false)
+}
+
+func (resolver *Resolver) QueryBlocks(ctx context.Context, id interface{}, params *jsonrpc.ParamsQueryBlocks, req *http.Request) jsonrpc.Response {
+	return resolver.handleMessage(ctx, id, jsonrpc.MethodQueryBlocks, params, req, false)
+}
+
+func (resolver *Resolver) SubmitTx(ctx context.Context, id interface{}, params *jsonrpc.ParamsSubmitTx, req *http.Request) jsonrpc.Response {
+	return resolver.handleMessage(ctx, id, jsonrpc.MethodSubmitTx, params, req, true)
+}
+
+func (resolver *Resolver) QueryTx(ctx context.Context, id interface{}, params *jsonrpc.ParamsQueryTx, req *http.Request) jsonrpc.Response {
+	return resolver.handleMessage(ctx, id, jsonrpc.MethodQueryTx, params, req, false)
+}
+
+func (resolver *Resolver) QueryPeers(ctx context.Context, id interface{}, params *jsonrpc.ParamsQueryPeers, req *http.Request) jsonrpc.Response {
+	return resolver.handleMessage(ctx, id, jsonrpc.MethodQueryPeers, params, req, false)
+}
+
+func (resolver *Resolver) QueryNumPeers(ctx context.Context, id interface{}, params *jsonrpc.ParamsQueryNumPeers, req *http.Request) jsonrpc.Response {
+	return resolver.handleMessage(ctx, id, jsonrpc.MethodQueryNumPeers, params, req, false)
+}
+
+func (resolver *Resolver) QueryEpoch(ctx context.Context, id interface{}, params *jsonrpc.ParamsQueryEpoch, req *http.Request) jsonrpc.Response {
+	return resolver.handleMessage(ctx, id, jsonrpc.MethodQueryEpoch, params, req, false)
+}
+
+func (resolver *Resolver) QueryStat(ctx context.Context, id interface{}, params *jsonrpc.ParamsQueryStat, req *http.Request) jsonrpc.Response {
+	return resolver.handleMessage(ctx, id, jsonrpc.MethodQueryStat, params, req, false)
+}
+
+func (resolver *Resolver) handleMessage(ctx context.Context, id interface{}, method string, params interface{}, r *http.Request, isSubmitTx bool) jsonrpc.Response {
+	query := r.URL.Query()
+	darknodeID := query.Get("id")
+	if darknodeID != "" {
+		if _, err := resolver.multiStore.Get(darknodeID); err != nil {
+			jsonErr := jsonrpc.NewError(jsonrpc.ErrorCodeInvalidParams, fmt.Sprintf("unknown darknode id %s", darknodeID), nil)
+			return jsonrpc.NewResponse(id, nil, &jsonErr)
+		}
+	}
+
+	reqWithResponder := lhttp.NewRequestWithResponder(ctx, id, method, params, query)
+	if isSubmitTx {
+		resolver.txCheckerRequests <- reqWithResponder
+	} else {
+		if ok := resolver.cacher.Send(reqWithResponder); !ok {
+			resolver.logger.Error("failed to send request to cacher, too much back pressure")
+			jsonErr := jsonrpc.NewError(jsonrpc.ErrorCodeInternal, "too much back pressure", nil)
+			return jsonrpc.NewResponse(id, nil, &jsonErr)
+		}
+	}
+
+	return <-reqWithResponder.Responder
```

Comment:
Maybe have a `select` statement from the `ctx.Done()` and the `reqWithResponder.Responder`.